# EDA of the data
* Performing an EDA of the data to discover preliminary patterns and problems which we will use in our preprocessing of the data
# Importing required libraries

In [5]:
import numpy as np
import pandas as pd

# Setting up display options for pandas

In [10]:
pd.set_option('display.max_columns', 500)
pd.set_option('display.max_rows', 500)

# Loading the data
* Try to load from pickle file as it is quicker to access
* If this fails, load the original CSV and save it to a pickle file
* <b>No point in doing anything with weather data as we only have 1 years worth of values so it is impossible to leverage with our chosen RNN/CNN based models to predict the next years worth of energy readings</b>
    * <b>Unless we leveraged them to produce multivariate predictions and also predicting the weathers values in order to use those as input for subsequent days: but this will likely result in worse performing energy predictions</b>
* Hence we are only looking into addInfo and consumption

In [9]:
#try to load pickle file as it is quicker
try:
    print("Loading consumption via pickle")
    df_energy = pd.read_pickle("../Data/Original_Data/consumption.pkl")
# if the file was not found then load the csv then save it to pickle for future loads
except:
    print("Loading consumption via csv")
    df_energy = pd.read_csv("../Data/Original_Data/consumption.csv")
    df_energy.to_pickle("../Data/Original_Data/consumption.pkl")

    #try to load pickle file as it is quicker
try:
    print("Loading addInfo via pickle")
    df_info = pd.read_pickle("../Data/Original_Data/addInfo.pkl")
# if the file was not found then load the csv then save it to pickle for future loads
except:
    print("Loading addInfo via csv")
    df_info = pd.read_csv("../Data/Original_Data/addInfo.csv")
    df_info.to_pickle("../Data/Original_Data/addInfo.pkl")

Loading consumption via pickle
Loading addInfo via pickle


# Inspecting df_energy and df_info

In [15]:
# inspecting the values of the dataframe
df_energy

meter_id  2017-01-01 00:00:00  \
0     0xa62b9f23553ff183f61e2bf943aab3d5983d02d7                  NaN   
1     0x459c834d1f6cfb5b734b82aa9f5410fa97fb70da                  NaN   
2     0x4a1ed36825360a058cec2bdd409fc2459e1ce54f                  NaN   
3     0x5b76d3c0e0aefc6e0a8d1d031f96388a23263407                  NaN   
4     0x943ebe39ef2be6ef807c42c5a647e27112ca5b0f                  NaN   
...                                          ...                  ...   
3243  0x7dd7a7b8ee1bec7c44b24f738c752482f6161065                0.011   
3244  0xfdaf9f857621ec06f2cf801f42a020a322835090                0.015   
3245  0xd28f2f001e0cd4d6c121a3cb2e1427207e170e18                0.222   
3246  0x47218b46abb2fcaade487a211911406dc6e13730                0.195   
3247  0xcd19e6fe3d887bc5dcac7ca18d46199695463fdd                0.023   

      2017-01-01 00:30:00  2017-01-01 01:00:00  2017-01-01 01:30:00  \
0                     NaN                  NaN                  NaN   
1                     NaN                  NaN                  NaN   
2                     NaN                  NaN                  NaN   
3                     NaN                  NaN                  NaN   
4                     NaN                  NaN                  NaN   
...                   ...                  ...                  ...   
3243                1.089                0.014                0.204   
3244                1.523                0.105                0.296   
3245                0.276                0.097                0.058   
3246                2.881                1.765                1.391   
3247                2.307                1.403                1.471   

      2017-01-01 02:00:00  2017-01-01 02:30:00  2017-01-01 03:00:00  \
0                     NaN                  NaN                  NaN   
1                     NaN                  NaN                  NaN   
2                     NaN                  NaN                  NaN   
3                     NaN                  NaN                  NaN   
4                     NaN                  NaN                  NaN   
...                   ...                  ...                  ...   
3243                0.015                0.159                0.014   
3244                0.067                0.171                0.025   
3245                0.045                0.045                0.038   
3246                1.609                1.362                1.352   
3247                1.421                1.305                1.299   

      2017-01-01 03:30:00  2017-01-01 04:00:00  2017-01-01 04:30:00  \
0                     NaN                  NaN                  NaN   
1                     NaN                  NaN                  NaN   
2                     NaN                  NaN                  NaN   
3                     NaN                  NaN                  NaN   
4                     NaN                  NaN                  NaN   
...                   ...                  ...                  ...   
3243                0.013                0.174                0.013   
3244                0.260                0.013                0.024   
3245                0.034                0.043                0.044   
3246                1.321                0.701                0.784   
3247                0.657                0.054                0.122   

      2017-01-01 05:00:00  2017-01-01 05:30:00  2017-01-01 06:00:00  \
0                     NaN                  NaN                  NaN   
1                     NaN                  NaN                  NaN   
2                     NaN                  NaN                  NaN   
3                     NaN                  NaN                  NaN   
4                     NaN                  NaN                  NaN   
...                   ...                  ...                  ...   
3243                0.014                0.014                0.014   
3244                0.012           

In [16]:
# inspecting the shape of the dataframe
df_energy.shape

(3248, 17521)

In [17]:
# inspecting the values of the dataframe
df_info

meter_id        dwelling_type  \
0     0x02b7418b821b2edeeb2baa23ba1580566b4b0e43  semi_detached_house   
1     0xaf32b2b497e9eb9c3f3de907aedc6a7914a11f41       terraced_house   
2     0xdd45ab28a7a40eadc04607d6da42e5ad923b1339  semi_detached_house   
3     0x2b538a5ebfc3780aef25227267063cb6410d943b       detached_house   
4     0xee032ed498c9e99a4c65e0f39896fb30e513c6b7                  NaN   
...                                          ...                  ...   
2138  0xb9c600842cb2d195d461d675f2758ddbf5bb45bb       terraced_house   
2139  0x429e76a41f3141d9e62c4975e6b5df303cb45e64             bungalow   
2140  0x65fbd9ba598e703d742a32058ea2b323fe4c1fca                  NaN   
2141  0x101f2dc409b3b95d2bb08ebe5d282f54317e3f8c             bungalow   
2142  0x9ad41b18c2be3a845acb599af1dc4570feb88489             bungalow   

      num_occupants  num_bedrooms heating_fuel hot_water_fuel boiler_age  \
0               NaN           3.0          NaN            NaN        NaN   
1               NaN           2.0          NaN            NaN        NaN   
2               NaN           NaN          NaN            NaN        NaN   
3               NaN           4.0          NaN            NaN        NaN   
4               NaN           NaN          NaN            NaN        NaN   
...             ...           ...          ...            ...        ...   
2138            2.0           2.0          gas            gas        old   
2139            NaN           3.0          NaN            NaN        NaN   
2140            NaN           NaN          NaN            NaN        NaN   
2141            NaN           4.0          NaN            NaN        NaN   
2142            NaN           3.0          NaN            NaN        NaN   

     loft_insulation wall_insulation heating_temperature  \
0                NaN             NaN                 NaN   
1                NaN             NaN                 NaN   
2                NaN             NaN                 NaN   
3                NaN             NaN                 NaN   
4                NaN             NaN                 NaN   
...              ...             ...                 ...   
2138               y               n            18_to_20   
2139             NaN             NaN                 NaN   
2140             NaN             NaN                 NaN   
2141             NaN             NaN                 NaN   
2142             NaN             NaN                 NaN   

     efficient_lighting_percentage  dishwasher  freezer  fridge_freezer  \
0                              NaN         NaN      NaN             NaN   
1                              NaN         NaN      NaN             NaN   
2                              NaN         NaN      NaN             NaN   
3                              NaN         NaN      NaN             NaN   
4                              NaN         NaN      NaN             NaN   
...                            ...         ...      ...             ...   
2138                     75_to_100         0.0      0.0             1.0   
2139                           NaN         NaN      NaN             NaN   
2140                           NaN         NaN      NaN             NaN   
2141                           NaN         NaN      NaN             NaN   
2142                           NaN         NaN      NaN             NaN   

      refrigerator  tumble_dryer  washing_machine  game_console  laptop   pc  \
0              NaN           NaN              NaN           NaN     NaN  NaN   
1              NaN           NaN              NaN           NaN     NaN  NaN   
2              NaN           NaN              NaN           NaN     NaN  NaN   
3              NaN           NaN              NaN           NaN     NaN  NaN   
4              NaN           NaN              NaN           NaN     NaN  NaN   
...            ...           ...              ...           ...     ...  ...   
2138           0.0           0.0              1.0           1.0     

In [18]:
# inspecting the shape of the dataframe
df_info.shape

(2143, 24)

* Can see that df_info only has entries for 2140 of the 3248 possible rows
* Must add the missing rows filled with NaN

# Adding the missing rows to addInfo

In [25]:
# first making both dataframes alphabetical in order of meter_id for ease of comparison
df_info.sort_values('meter_id')

meter_id        dwelling_type  \
750   0x000f697092696c27b0bb489ea4c11280ef72ab82       detached_house   
88    0x001d6227832325ab6167b82d6d7175d254f57c3a  semi_detached_house   
1643  0x005958406351bb29580475df698b5f1070096397       detached_house   
694   0x00735e1aa83d97d9209cce47d8777b4ff3cc0554       terraced_house   
2056  0x0079fa6c2cbd3de97c8aa6073ff365b41ee087ab       terraced_house   
...                                          ...                  ...   
1922  0xffb4c5014834a46d09d0092e748c1ef5acd8733f  semi_detached_house   
1463  0xffbdc1d9209f45668fb0be65b6d50cb999890ccb                  NaN   
1685  0xffcbcdc7c3ce3451252b84c2aaf6f40d652d16d5       detached_house   
441   0xffe91f021d145e1560fcad8c3af5ac0ef9d76a02                  NaN   
2119  0xfff895258c21f1a58fc06538173d02b621021ad4                 flat   

      num_occupants  num_bedrooms heating_fuel hot_water_fuel boiler_age  \
750             NaN           4.0          NaN            NaN        NaN   
88              NaN           3.0          NaN            NaN        NaN   
1643            NaN           4.0          NaN            NaN        NaN   
694             NaN           3.0          NaN            NaN        NaN   
2056            NaN           3.0          NaN            NaN        NaN   
...             ...           ...          ...            ...        ...   
1922            NaN           3.0          NaN            NaN        NaN   
1463            NaN           3.0          NaN            NaN        NaN   
1685            NaN           3.0          NaN            NaN        NaN   
441             NaN           3.0          NaN            NaN        NaN   
2119            NaN           3.0          NaN            NaN        NaN   

     loft_insulation wall_insulation heating_temperature  \
750              NaN             NaN                 NaN   
88               NaN             NaN                 NaN   
1643             NaN             NaN                 NaN   
694              NaN             NaN                 NaN   
2056             NaN             NaN                 NaN   
...              ...             ...                 ...   
1922             NaN             NaN                 NaN   
1463             NaN             NaN                 NaN   
1685             NaN             NaN                 NaN   
441              NaN             NaN                 NaN   
2119             NaN             NaN                 NaN   

     efficient_lighting_percentage  dishwasher  freezer  fridge_freezer  \
750                            NaN         NaN      NaN             NaN   
88                             NaN         NaN      NaN             NaN   
1643                           NaN         NaN      NaN             NaN   
694                            NaN         NaN      NaN             NaN   
2056                           NaN         NaN      NaN             NaN   
...                            ...         ...      ...             ...   
1922                           NaN         NaN      NaN             NaN   
1463                           NaN         NaN      NaN             NaN   
1685                           NaN         NaN      NaN             NaN   
441                            NaN         NaN      NaN             NaN   
2119                           NaN         NaN      NaN             NaN   

      refrigerator  tumble_dryer  washing_machine  game_console  laptop  pc  \
750            NaN           NaN              NaN           NaN     NaN NaN   
88             NaN           NaN              NaN           NaN     NaN NaN   
1643           NaN           NaN              NaN           NaN     NaN NaN   
694            NaN           NaN              NaN           NaN     NaN NaN   
2056           NaN           NaN              NaN           NaN     NaN NaN   
...            ...           ...              ...           ...     ...  ..   
1922           NaN           NaN              NaN           NaN     NaN NaN

In [26]:
df_energy.sort_values('meter_id')

meter_id  2017-01-01 00:00:00  \
1124  0x0001f1c389823f953b2eaee0a61c33539744da0c                  NaN   
602   0x000f697092696c27b0bb489ea4c11280ef72ab82                  NaN   
972   0x0019979ee12c59accd24d1c83291528ced1bb734                  NaN   
356   0x001d6227832325ab6167b82d6d7175d254f57c3a                  NaN   
133   0x005958406351bb29580475df698b5f1070096397                  NaN   
...                                          ...                  ...   
1231  0xffb4c5014834a46d09d0092e748c1ef5acd8733f                  NaN   
2580  0xffbdc1d9209f45668fb0be65b6d50cb999890ccb                  NaN   
2198  0xffcbcdc7c3ce3451252b84c2aaf6f40d652d16d5                  NaN   
2792  0xffe91f021d145e1560fcad8c3af5ac0ef9d76a02                  NaN   
3098  0xfff895258c21f1a58fc06538173d02b621021ad4                0.094   

      2017-01-01 00:30:00  2017-01-01 01:00:00  2017-01-01 01:30:00  \
1124                  NaN                  NaN                  NaN   
602                   NaN                  NaN                  NaN   
972                   NaN                  NaN                  NaN   
356                   NaN                  NaN                  NaN   
133                   NaN                  NaN                  NaN   
...                   ...                  ...                  ...   
1231                  NaN                  NaN                  NaN   
2580                  NaN                  NaN                  NaN   
2198                  NaN                  NaN                  NaN   
2792                  NaN                  NaN                  NaN   
3098                0.093                0.094                0.089   

      2017-01-01 02:00:00  2017-01-01 02:30:00  2017-01-01 03:00:00  \
1124                  NaN                  NaN                  NaN   
602                   NaN                  NaN                  NaN   
972                   NaN                  NaN                  NaN   
356                   NaN                  NaN                  NaN   
133                   NaN                  NaN                  NaN   
...                   ...                  ...                  ...   
1231                  NaN                  NaN                  NaN   
2580                  NaN                  NaN                  NaN   
2198                  NaN                  NaN                  NaN   
2792                  NaN                  NaN                  NaN   
3098                0.091                  0.1                 0.09   

      2017-01-01 03:30:00  2017-01-01 04:00:00  2017-01-01 04:30:00  \
1124                  NaN                  NaN                  NaN   
602                   NaN                  NaN                  NaN   
972                   NaN                  NaN                  NaN   
356                   NaN                  NaN                  NaN   
133                   NaN                  NaN                  NaN   
...                   ...                  ...                  ...   
1231                  NaN                  NaN                  NaN   
2580                  NaN                  NaN                  NaN   
2198                  NaN                  NaN                  NaN   
2792                  NaN                  NaN                  NaN   
3098                0.088                0.088                0.099   

      2017-01-01 05:00:00  2017-01-01 05:30:00  2017-01-01 06:00:00  \
1124                  NaN                  NaN                  NaN   
602                   NaN                  NaN                  NaN   
972                   NaN                  NaN                  NaN   
356                   NaN                  NaN                  NaN   
133                   NaN                  NaN                  NaN   
...                   ...                  ...                  ...   
1231                  NaN                  NaN                  NaN   
2580                  NaN           

In [34]:
#adding the missing meter rows to df_info
df_info_all_meters = pd.merge(df_info, df_energy["meter_id"], on="meter_id",how="outer")

In [35]:
#checking the values are correct now
df_info_all_meters.sort_values('meter_id')

meter_id        dwelling_type  \
2532  0x0001f1c389823f953b2eaee0a61c33539744da0c                  NaN   
750   0x000f697092696c27b0bb489ea4c11280ef72ab82       detached_house   
2468  0x0019979ee12c59accd24d1c83291528ced1bb734                  NaN   
88    0x001d6227832325ab6167b82d6d7175d254f57c3a  semi_detached_house   
1643  0x005958406351bb29580475df698b5f1070096397       detached_house   
...                                          ...                  ...   
1922  0xffb4c5014834a46d09d0092e748c1ef5acd8733f  semi_detached_house   
1463  0xffbdc1d9209f45668fb0be65b6d50cb999890ccb                  NaN   
1685  0xffcbcdc7c3ce3451252b84c2aaf6f40d652d16d5       detached_house   
441   0xffe91f021d145e1560fcad8c3af5ac0ef9d76a02                  NaN   
2119  0xfff895258c21f1a58fc06538173d02b621021ad4                 flat   

      num_occupants  num_bedrooms heating_fuel hot_water_fuel boiler_age  \
2532            NaN           NaN          NaN            NaN        NaN   
750             NaN           4.0          NaN            NaN        NaN   
2468            NaN           NaN          NaN            NaN        NaN   
88              NaN           3.0          NaN            NaN        NaN   
1643            NaN           4.0          NaN            NaN        NaN   
...             ...           ...          ...            ...        ...   
1922            NaN           3.0          NaN            NaN        NaN   
1463            NaN           3.0          NaN            NaN        NaN   
1685            NaN           3.0          NaN            NaN        NaN   
441             NaN           3.0          NaN            NaN        NaN   
2119            NaN           3.0          NaN            NaN        NaN   

     loft_insulation wall_insulation heating_temperature  \
2532             NaN             NaN                 NaN   
750              NaN             NaN                 NaN   
2468             NaN             NaN                 NaN   
88               NaN             NaN                 NaN   
1643             NaN             NaN                 NaN   
...              ...             ...                 ...   
1922             NaN             NaN                 NaN   
1463             NaN             NaN                 NaN   
1685             NaN             NaN                 NaN   
441              NaN             NaN                 NaN   
2119             NaN             NaN                 NaN   

     efficient_lighting_percentage  dishwasher  freezer  fridge_freezer  \
2532                           NaN         NaN      NaN             NaN   
750                            NaN         NaN      NaN             NaN   
2468                           NaN         NaN      NaN             NaN   
88                             NaN         NaN      NaN             NaN   
1643                           NaN         NaN      NaN             NaN   
...                            ...         ...      ...             ...   
1922                           NaN         NaN      NaN             NaN   
1463                           NaN         NaN      NaN             NaN   
1685                           NaN         NaN      NaN             NaN   
441                            NaN         NaN      NaN             NaN   
2119                           NaN         NaN      NaN             NaN   

      refrigerator  tumble_dryer  washing_machine  game_console  laptop  pc  \
2532           NaN           NaN              NaN           NaN     NaN NaN   
750            NaN           NaN              NaN           NaN     NaN NaN   
2468           NaN           NaN              NaN           NaN     NaN NaN   
88             NaN           NaN              NaN           NaN     NaN NaN   
1643           NaN           NaN              NaN           NaN     NaN NaN   
...            ...           ...              ...           ...     ...  ..   
1922           NaN           NaN              NaN           NaN     NaN NaN

In [36]:
#checking the shape is correct now
df_info_all_meters.shape

(3248, 24)

As can be seen we have now added the missing rows to df_info

# EDA on AddInfo

In [37]:
# inspecting the types of each column
df_info_all_meters.dtypes

meter_id                          object
dwelling_type                     object
num_occupants                    float64
num_bedrooms                     float64
heating_fuel                      object
hot_water_fuel                    object
boiler_age                        object
loft_insulation                   object
wall_insulation                   object
heating_temperature               object
efficient_lighting_percentage     object
dishwasher                       float64
freezer                          float64
fridge_freezer                   float64
refrigerator                     float64
tumble_dryer                     float64
washing_machine                  float64
game_console                     float64
laptop                           float64
pc                               float64
router                           float64
set_top_box                      float64
tablet                           float64
tv                               float64
dtype: object

In [41]:
#looking at the unique values for each column
for column in df_info_all_meters.drop(["meter_id"],axis=1).columns:
    print(column+" unique value count: "+str(df_info_all_meters[column].nunique()))
    print(df_info_all_meters[column].unique())
    print()

dwelling_type unique value count: 5
['semi_detached_house' 'terraced_house' 'detached_house' nan 'bungalow'
 'flat']

num_occupants unique value count: 4
[nan  2.  4.  3.  1.]

num_bedrooms unique value count: 5
[ 3.  2. nan  4.  5.  1.]

heating_fuel unique value count: 4
[nan 'gas' 'elec' 'other' 'lpg_oil']

hot_water_fuel unique value count: 3
[nan 'gas' 'elec' 'other']

boiler_age unique value count: 2
[nan 'new' 'old']

loft_insulation unique value count: 2
[nan 'y' 'n']

wall_insulation unique value count: 5
[nan 'y_cavity' 'n' 'y_internal' 'not_sure' 'y_external']

heating_temperature unique value count: 4
[nan '18_to_20' 'above_20' 'below_18' 'not_sure']

efficient_lighting_percentage unique value count: 4
[nan '0_to_25' '25_to_50' '50_to_75' '75_to_100']

dishwasher unique value count: 3
[nan  1.  0.  2.]

freezer unique value count: 3
[nan  2.  1.  0.]

fridge_freezer unique value count: 3
[nan  2.  0.  1.]

refrigerator unique value count: 3
[nan  2.  1.  0.]

tumble_dryer u